lib/stores/S3Store.js

```diff
@@ -276,57 +286,96 @@ class S3Store extends DataStore {
-                            });
+   * uploads a stream to s3 using multiple parts
+   *
+   * @param {Object}                metadata            upload metadata
+   * @param {http<IncomingMessage>} req                 incoming request
+   * @param {Number}                currentPartNumber number of the current part/chunk
+   * @return {Promise<Number>}                         which resolves with the current offset
+   * @memberof S3Store
+   */
+    _processUpload(metadata, req, currentPartNumber) {
+        return new Promise((resolve, reject) => {
+            const splitter = new FileStreamSplitter(
+                {},
+                { maxChunkSize: this.part_size, directory: '.' }
+            );
+
+            const promises = [];
+
+            let pendingChunkFilepath = null;
+
+            stream.pipeline(req, splitter, (pipelineErr) => {
+                if (pipelineErr) {
+                    if (pendingChunkFilepath !== null) {
+                        fs.rm(pendingChunkFilepath, (fileRemoveErr) => {
+                            if (fileRemoveErr) {
+                                log(
+                                    `[${metadata.file.id}] failed to remove chunk ${pendingChunkFilepath}`
+                                );
+                            }
+                        });
+                    }
+                }
 
-                            this._clearCache(metadata.file.id);
+                promises.push(pipelineErr ? Promise.reject(pipelineErr) : Promise.resolve());
+                resolve(promises);
+            });
 
-                            return current_offset.size;
-                        })
-                        .catch((err) => {
-                            throw err;
+            splitter.on('chunkStarted', (filepath) => {
+                pendingChunkFilepath = filepath;
+            });
+
+            splitter.on('chunkFinished', (filepath) => {
+                pendingChunkFilepath = null;
+
+                const partNumber = currentPartNumber++;
+
+                const p = Promise.resolve()
+                    // TODO: we don't have headers here to check the offset and upload-length
+                    // to know whether this `isFinalRequest`. Find a different way to check
+                    // the part size on the final request.
+                    //
+                    // .then(async() => {
+                    //     const stats = await fs.promises.stat(filepath);
+                    //     if (!isFinalRequest) {
+                    //         // We could modify chunk size in a way that last part is as small as possible.
+                    //         // Ideally we would not discard any data.
+                    //         if (stats.size < 5 * 1024 * 1024) {
+                    //             throw new Error('Part too small');
+                    //         }
+                    //     }
+                    //     return undefined;
+                    // })
```

Comment:
@mitjap I had to remove this check. You were accessing headers from `req` but in all stores and tests `req` is a stream, not a http request, so the headers don't exist. Need to find a different way to handle this

Reply:
I was initially implementing an alternative way to check this and error but then then I realized this is redundant. It should never happen that a chunk is smaller than 5mb if it's not the final chunk (if a file is bigger than 5mb) and we could test that. I also don't think tusd errors on this?

Perhaps related:
https://github.com/tus/tusd/blob/d72854fa048e314a86b60cad98c8144341f297b5/pkg/s3store/s3store.go#L708-L717